pulsar-core/src/pdk/module.rs

```diff
@@ -160,28 +160,65 @@ pub type ErrorSender = mpsc::Sender<ModuleError>;
 pub type ModuleError = Box<dyn std::error::Error + Send + Sync + 'static>;
 
 impl ModuleSender {
+    /// Send an event to the [`Bus`].
     pub fn send(&self, process: Pid, timestamp: Timestamp, payload: Payload) {
-        self.send_internal(process, timestamp, payload, false)
+        self.send_internal(process, timestamp, payload, None)
     }
 
-    pub fn send_threat(&self, process: Pid, timestamp: Timestamp, payload: Payload) {
-        self.send_internal(process, timestamp, payload, true)
+    /// Send a threat which was caused by another event to the [`Bus`].
+    /// The new event shares the source headers, but has a new payload.
+    pub fn send_derived(&self, source: &Event, payload: Payload) {
+        let header = source.header.clone();
+        let _ = self.tx.send(Event { header, payload });
+    }
+
+    /// Send a threat to the [`Bus`].
+    pub fn send_threat(&self, process: Pid, timestamp: Timestamp, info: impl Into<Value>) {
+        self.send_internal(process, timestamp, Payload::Empty, Some(info.into()))
+    }
+
+    /// Send an event which was caused by another one to the [`Bus`].
+    /// The new event shares the source headers, but has a new payload.
+    pub fn send_threat_derived(&self, source_event: &Event, info: impl Into<Value>) {
+        let threat = Threat {
+            source: self.module_name.clone(),
+            info: info.into(),
+        };
+
+        let _ = self.tx.send(Event {
+            header: Header {
+                threat: Some(threat),
+                ..source_event.header.clone()
+            },
+            payload: source_event.payload.clone(),
+        });
     }
 
-    /// Send a [`Payload`] to the [`Bus`].
-    fn send_internal(&self, process: Pid, timestamp: Timestamp, payload: Payload, is_threat: bool) {
+    /// Send a [`Payload`] to the [`Bus`] with optional [`Value`].
+    /// Adding a `Value` fills the [`Threat`] field in the header.
+    fn send_internal(
+        &self,
+        process: Pid,
+        timestamp: Timestamp,
+        payload: Payload,
+        info: Option<Value>,
+    ) {
         let tx = self.tx.clone();
-        let source = self.module_name.clone();
         let process_tracker = self.process_tracker.clone();
         let module_name = self.module_name.clone();
         tokio::spawn(async move {
+            let threat = info.map(|info| Threat {
+                source: module_name.clone(),
+                info,
+            });
+
             let mut header = Header {
-                source,
-                is_threat,
+                source: module_name.clone(),
```

Comment:
Not sure if this is a bug or if I got it wrong. I assumed that when sending derived events which just fills the threat option, we wouldn't change the source. For consistency, I would assume this to expand to all derived events.

Reply:
I dont't understand, when a send*derived method is used, it will not use the `send_internal`

Reply:
My bad, I wrongly assumed without checking that all 4 methods would call send_internal